src/viam/sdk/common/client_helper.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <grpcpp/client_context.h>
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+
+namespace viam {
+namespace sdk {
+
+template <typename ClientType, typename StubType, typename RequestType, typename ResponseType>
+class ClientHelper {
+    static void default_rsc_(RequestType&) {}
+    static void default_rhc_(const ResponseType&) {}
+    static void default_ehc_(const ::grpc::Status& status) {
+        throw std::runtime_error(status.error_message());
+    }
+
+   public:
+    using PFn = ::grpc::Status (StubType::*)(::grpc::ClientContext*,
+                                             const RequestType&,
+                                             ResponseType*);
+    explicit ClientHelper(ClientType* client, StubType* stub, PFn pfn)
+        : client_(client), stub_(stub), pfn_(pfn) {}
+
+    ClientHelper& with(const AttributeMap& extra) {
+        return with(extra, default_rsc_);
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(RequestSetupCallable&& rsc) {
+        std::forward<RequestSetupCallable>(rsc)(request_);
+        return *this;
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(const AttributeMap& extra, RequestSetupCallable&& rsc) {
+        *request_.mutable_extra() = map_to_struct(extra);
+        return with(std::forward<RequestSetupCallable>(rsc));
+    }
+
+    template <typename ResponseHandlerCallable = decltype(default_rhc_)>
+    auto invoke(ResponseHandlerCallable&& rhc = default_rhc_) {
+        return invoke(std::forward<ResponseHandlerCallable>(rhc), default_ehc_);
+    }
+
+    template <typename ResponseHandlerCallable, typename ErrorHandlerCallable>
+    auto invoke(ResponseHandlerCallable&& rhc, ErrorHandlerCallable&& ehc) {
+        *request_.mutable_name() = client_->name();
+        ::grpc::ClientContext ctx;
+        set_client_ctx_authority(ctx);
```

Comment:
Note that I don't think this obviates the need for #179 of some form. This class shouldn't be responsible for properly configuring a client context. But it does mean that if we convert the codebase over to `ClientHelper`, there will be fewer places where the factory or wrapper class needs to be injected.

Reply:
> This class shouldn't be responsible for properly configuring a client context.

If we do convert the whole codebase to `ClientHelper` (including the complex module example), why would we need to configure a client context anywhere else?

Reply:
Cool that makes sense; thanks!